addons/udes_advanced_packaging/models/stock_picking.py

```diff
@@ -0,0 +1,156 @@
+        """
+        If the picking type of the picking in self has full package
+        reservation enabled, whole packages will be reserved.
+        """
+        Quant = self.env["stock.quant"]
+        MoveLine = self.env["stock.move.line"]
+
+        # do not reserve full packages when bypass_reserve_full packages
+        # is set in the context as True
+        if not self.env.context.get("bypass_reserve_full_packages"):
+            for picking in self.filtered(lambda p: p._should_reserve_full_packages()):
+                all_quants = Quant.browse()
+                remaining_qtys = defaultdict(int)
+
+                # get all packages
+                packages = picking.mapped("move_line_ids.package_id")
+                for package in packages:
+                    move_lines = picking.mapped("move_line_ids").filtered(
+                        lambda ml: ml.package_id == package
+                    )
+                    pack_products = frozenset(package._get_all_products_quantities().items())
+                    mls_products = frozenset(move_lines._get_all_products_quantities().items())
+                    if pack_products != mls_products:
+                        # move_lines do not match the quants
+                        pack_mls = MoveLine.search(
+                            [
+                                ("package_id", "child_of", package.id),
+                                ("state", "not in", ["done", "cancel"]),
+                            ]
+                        )
+                        other_pickings = pack_mls.mapped("picking_id") - picking
+                        if other_pickings:
+                            raise ValidationError(
+                                _("The package is reserved in other pickings: %s")
+                                % ",".join(other_pickings.mapped("name"))
+                            )
+
+                        quants = package._get_contained_quants()
+                        all_quants |= quants
+                        for product, qty in quants.group_quantity_by_product(
+                            only_available=True
+                        ).items():
+                            remaining_qtys[product] += qty
+                if remaining_qtys:
+                    # Context variables:
+                    # - filter the quants used in _create_moves() to be
+                    # the ones of the packages to be completed
+                    # - add bypass_reserve_full_packages at the context
+                    # to avoid to be called again inside _create_moves()
+                    picking.with_context(
+                        bypass_reserve_full_packages=True, quant_ids=all_quants.ids
+                    )._create_moves(
+                        remaining_qtys,
+                        values={"priority": picking.priority},
+                        confirm=True,
+                        assign=True,
+                    )
+
+    def _create_moves(
```

Comment:
We should use the create_moves behaviour in udes_stock here i believe

Reply:
prepare moves, then pass to  create move

Reply:
Do you mean to copy this method into udes_stock?

Reply:
No it already exists i believe

Reply:
I probably need to rebase then lol

Reply:
No i think _create_moves does not exist with the same name, but create_move or creete_moves does, you should re-use that functionality

Reply:
This should be sorted now, as the existing helper method in `udes_stock` is now being used.